### PR TITLE
pkg/apparmor: scratch documentation of apparmor

### PR DESCRIPTION
**Pull request by zyga**
We're currently growing support for apparmor 2.13+ and I was thinking
about how we structure the code related to that. I'm worried of the
growing complexity in interfaces/apparmor (and release/apparmor.go) and
would like to move that code to pkg/apparmor where it can be referenced,
as appropriate from original locations. In addition I'd like to clearly
separate old and new apparmor, using the chance to document how it
behaves in each version.

This patch is by no means complete but I'd like to use it to start a
discussion about moving the code out, gardening it a little bit and
dropping it here. I chose the directory pkg/apparmor as we can expect
similar treatment to seccomp code without adding a new top-level element
for each package and the pkg/foo naming scheme seems to be common in the
go ecosystem.

Signed-off-by: Zygmunt Krynicki <zygmunt.krynicki@canonical.com>

### CODE REVIEW

**Comment by pedronis** [comment]
There is no general agreement about the pkg/ style, I just saw over the weekend people discussing exactly that, I find it a bit low-content as naming style, it would also make us question whether to move more stuff.

**Comment by zyga** [comment]
The name `pkg` is nice and short. It's used by many go projects though that is not any general endorsement across the ecosystem. To me it looks like a `src` or `include` directory and allows us to stash library code away from the, already very busy, top level directory.

This code could, in theory, live in `interfaces/apparmor` but that location is strictly about *interfaces*. We already have a similar problem where, for totally historic reasons, apparmor code also lives in the `release` package.

**Comment by pedronis** [comment]
I think sandbox/apparmor sandbox/seccomp would follow more our current organisational style (which I don't see pressure to change)

**Comment by zyga** [comment]
I've renamed the package to `sandbox/apparmor`.  Depending on feedback I could help move things from the various places that "do" apparmor there.

File: sandbox/apparmor/legacy.go [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+// Source profiles are typically stored as files in /etc/apparmor.d/.
+// The file name is the absolute path of the executable with forward slashes
+// replaced with dots. For example the file /etc/apparmor.d/usr.bin.man
+// contains a profile that attaches to /usr/bin/man. In addition the same file
+// defines additional profiles man_groff and man_filter, that various
+// executables started from /usr/bin/man assume.
+//
+// Source profiles routinely include definitions from
+// /etc/apparmor.d/abstractions as well as tunables that can be adjusted by
+// local administrator, from /etc/apparmor.d/tunables.
```
Comment (jdstrand) [inline sandbox/apparmor/legacy.go]: Again, this is default behavior but entirely dependent on what the profiles use. Eg Ubuntu Touch would #include files from /usr and the parser supports specifying a different base directory with --base.

File: sandbox/apparmor/legacy.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+//
+// Once compiled, binary profiles are typically cached as files in
+// /etc/apparmor.d/cache. The cache is based on mtime of the source profile and
+// of the cached profile. The cache is unaware of the kernel feature set.  This
```
Comment (jdstrand) [inline sandbox/apparmor/legacy.go]: That's not true. /etc/apparmor.d/cache/.features is present prior to 2.13 and was used as part of the parser's decision to invalidate the cache.

File: sandbox/apparmor/legacy.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+//
+// Binary profiles have the same file name as the source profile.
+// Binary profiles, just like source profiles, can contain multiple actual
+// profiles. Binary profile can be directly loaded into the kernel without any
+// extra processing.
+//
+// Apart from the aforementioned directories in /etc/apparmor.d/, at least on
+// some distributions, there are more binary profiles present in
+// /var/cache/apparmor and more source profiles present in
+// /var/lib/snapd/apparmor/profiles. Profiles stored there don't follow the
```
Comment (jdstrand) [inline sandbox/apparmor/legacy.go]: Or other places... see above.

File: sandbox/apparmor/modern.go [hunk 4]
```diff
@@ -0,0 +1,7 @@
+package apparmor
+
+// modern represents apparmor user-space version since 1.13
```
Comment (jdstrand) [inline sandbox/apparmor/modern.go]: 2.13*

**Comment by zyga** [comment]
@jdstrand thank you for all the comments. I will adjust this for another pass. I think some explanation is in order. I was looking at the 2.13 PR and was somewhat uncomfortable with clustering of apparmor specific logic 2.12 vs 2.13 along with snapd logic. I though it would be nice to abstract apparmor differences behind a new module, taking the opportunity to document how 2.12 and 2.13 differ.

**Comment by pedronis** [comment]
Legacy and modern are not clear names,  Pre2_13  and 21_3Plus or something. Where/how do we pick one or the other?

Also is still a bit unclear to me there is enough actual difference to warrant a full new package, vs a helper.go file in interfaces/apparmor.

**Comment by zyga** [comment]
I will return here tomorrow.

**Comment by pedronis** [comment]
I'm closing this, is still a bit unclear if it's a win,  and I don't think we have overall bandwidth ATM to work/think  through it. It can be reopened later.

**Pull request closed without merging**